src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private Node<T> next;
+        private Node<T> prev;
+        private T value;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail == null) {
+            head = newNode;
+        }
+        if (tail != null) {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index is not valid :" + index);
+        }
+        if (head == null && index == 0) {
+            head = newNode;
+            tail = newNode;
+        }
+        if (index == 0 && head != null) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        }
+        if (index == size) {
+            newNode.prev = tail;
+            if (tail != null) {
+                tail.next = newNode;
+            }
+            tail = newNode;
+        }
+        if (index > 0 && index < size) {
+            Node<T> current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            newNode.prev = current.prev;
+            newNode.next = current;
+            if (current.prev != null) {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index is not valid :" + index);
+        }
+        Node<T> current = head;
```

Comment:
let's create method findNodeByIndex() and reuse it in get, set and remove methods